### ISSUE TITLE
UIA improvement ideas

### ISSUE THREAD

**Issue by Sorunome** [matrix-org/matrix-doc#2408]
So, soru worked a lot with UIA auth recently and has a few ideas how to improve it. Not quite sure
if to make a full MSC about this, and wanting to write down the ideas in a more permanent way than
just a chat, thus this issue.

The problems soru sees are the following:
 - optional stages have to be defined with multiple flows
 - as soon as you have more than one optional stage it gets hard (as you'd need multiple m.login.dummy)
 - also with optional stages the amount of flows grows quickly
 - the server can't later on add a stage, which would e.g. be needed for 2fa (server checks after
   m.login.password if the user has enable 2fa and forces based on that)
 - The "fix" for the above to adding an extra roundtrip seems.....hacky
 - the last request of UIA is the same request as the underlying endpoint, causing confusion for if
   the returned error is due to UIA or due to the underlying endpoint

Fix ideas:
 - Instead of the server presenting the flows the client can complete, it should say what the next
   possible stages are. The `completed` array is obsolete. This would easily fix the issue with
   optional steps, as it can present two stages, like the email stage and `m.login.dummy`.  
   This would also fix the issue with 2fa mentioned above: after an `m.login.password` the server
   notices that the user has 2fa enabled and then forces them to complete the 2fa stage next.  
   This also means that the server only needs to reply the `params` of the current available stages
 - Introduce `m.login.completed` as another dummy-stage to indicate the completion of UIA, mitigating
   the problem of the the ambiguity if the error comes from UIA or the underlying endpoint

**Comment by joepie91** [comment]
> Instead of the server presenting the flows the client can complete, it should say what the next
possible stages are.

One problem I can see with this, is that some clients may want to know what to expect after completing a particular stage, so you'd probably still want to provide a list of possible flows (with maybe an 'optional' marker), just so that clients for which this is relevant can still use that information.

(Think a client that might want to show a choice between "e-mail link + SMS code" and "username/password".)

**Comment by Sorunome** [comment]
> One problem I can see with this, is that some clients may want to know what to expect after completing a particular stage, so you'd probably still want to provide a list of possible flows (with maybe an 'optional' marker), just so that clients for which this is relevant can still use that information.

And then, in the case of 2fa, that one just isn't advocated in that array and you encounter it once the server says that is the next stage you have to complete? So that is more of a "this is what the flows *could* look like"?

**Comment by joepie91** [comment]
> And then, in the case of 2fa, that one just isn't advocated in that array and you encounter it once the server says that is the next stage you have to complete? So that is more of a "this is what the flows _could_ look like"?

Right, exactly, it would be purely "advisory" / "for illustrative purposes" / whatever you want to call it; the actual following of the flow would be determined by the next-stage hint.

Ideally it would still list the 2FA in that API as an *optional* step, though, to support "step 1 of 3" type UIs (where it's okay if you suddenly skip a step, but users might get frustrated if there's a step *more* than they expected).
